### PR TITLE
Add LinkGraph interface

### PR DESCRIPTION
**Pull request by laughedelic**
A graph that connects two explicitly defined graphs:
- can have its own vertices/edges
- ~~can~~ should have edges between vertices of those two graphs

### CODE REVIEW

**Comment by laughedelic** [comment]
This is basically a copy-paste from https://github.com/bio4j/bio4j/pull/150. I don't really like this solution because it duplicates a big piece of code related to the edge type initialization and its indexes, plus changes access modifiers in `TypedGraph` (those mutable values should be private). See my comment above about slightly different solution.

@eparejatobes review this please.

File: src/main/java/com/bio4j/angulillos/TypedLinkGraph.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
+package com.bio4j.angulillos;
+
+import com.bio4j.angulillos.*;
+import java.util.Set;
```
Comment (laughedelic) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: Thank you @codacy-bot 😘

**Comment by laughedelic** [comment]
Travis failure is strange. Looks like it cannot resolve some sbt-related artifacts "/

File: src/main/java/com/bio4j/angulillos/TypedLinkGraph.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+
+      protected LinkEdgeType(
+        TypedGraph<SG,RV,RE>.VertexType<S> sourceType,
+        TypedGraph<TG,RV,RE>.VertexType<T> targetType
+      ) {
+
+        this.sourceType = sourceType;
+        this.targetType = targetType;
+
+        sourceType.outEdges.add( self() );
```
Comment (eparejatobes) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: what was the point of these in/outEdges sets?

File: src/main/java/com/bio4j/angulillos/TypedLinkGraph.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+      private final TypedGraph<SG,RV,RE>.VertexType<S> sourceType;
+      private final TypedGraph<TG,RV,RE>.VertexType<T> targetType;
+
+      @Override public final TypedGraph<SG,RV,RE>.VertexType<S> sourceType() { return this.sourceType; }
+      @Override public final TypedGraph<TG,RV,RE>.VertexType<T> targetType() { return this.targetType; }
+
+      // NOTE: this initializer block will be inherited and will add each edge type to the set
+      {
+        if (
+          TypedLinkGraph.this.edgeTypes.removeIf( (AnyEdgeType et) ->
```
Comment (eparejatobes) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: I know this is not strictly part of this PR but why removeIf and not contains or something like it?

File: project/build.properties [hunk 4]
```diff
@@ -1 +1 @@
-sbt.version=0.13.11
+sbt.version=0.13.12
```
Comment (laughedelic) [inline project/build.properties]: Probably it was something different? I use it everywhere without problems.

File: src/main/java/com/bio4j/angulillos/TypedLinkGraph.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            et._label().equals( self()._label() )
+          )
+        ) {
+          throw new IllegalArgumentException("The graph contains duplicate edge type: " + self()._label());
+        }
+
+        TypedLinkGraph.this.edgeTypes.add(self());
+      }
+
+      protected LinkEdgeType(
```
Comment (laughedelic) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: I'm not sure, but I think init block has to be before the constructor. Also I prefer to define those fields/accessors before I use them in the constructor.

File: src/main/java/com/bio4j/angulillos/TypedLinkGraph.java (hunk 3 above)
Comment (laughedelic) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: Because `contains` take an exact object that you're looking for, while here we need to search by apredicate (comparing the labels).

File: src/main/java/com/bio4j/angulillos/TypedLinkGraph.java (hunk 2 above)
Comment (laughedelic) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: Similarly to the sets of "untyped" edges and vertices in the `TypedGraph`, each [vertex type](https://github.com/bio4j/angulillos/blob/master/src/main/java/com/bio4j/angulillos/AnyVertexType.java#L12-L13) has a set of the in/out edge types. It may be needed for the automatic schema creation (or depending on the backend it could be used for some advanced vertex type tuning).
Also I think we discussed this in the context of #63 for schema inspection.

File: src/main/java/com/bio4j/angulillos/TypedLinkGraph.java (hunk 3 above)
Comment (eparejatobes) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: sure, but you can `stream().filter(e -> whathever).findFirst()` for example

File: src/main/java/com/bio4j/angulillos/TypedLinkGraph.java (hunk 2 above)
Comment (eparejatobes) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: OK yeah, now I remember, thanks :+1:

**Comment by laughedelic** [comment]
Thanks. 

> About the alternative you mention, that means you don't need to duplicate the initializer blocks?

Yes, also constructors. The idea is just to have a common class with that code and two descendants: one with simple edges (only withtin the graph), another with arbitrary edges. Actually it's more like this:

- `GenericTypedGraph` with **protected** `GenericEdgeType` from any graph to any (like in the added here code)
  + `TypedSimpleGraph extends GenericTypedGraph` with `EdgeType` that fixes source/target graphs to the one where it's defined
    * `TypedLinkGraph extends TypedSimpleGraph` with both inherited simple `EdgeType` and added `LinkEdgeType` which just inherits `GenericEdgeType` without any restrictions.

File: src/main/java/com/bio4j/angulillos/TypedLinkGraph.java (hunk 3 above)
Comment (laughedelic) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: I think `removeIf` is more straightforward. If we find (remove) one, it throws _an exception_ anyway 👍

File: project/build.properties (hunk 4 above)
Comment (laughedelic) [inline project/build.properties]: I remember now, it was related to nice-sbt-settings at some point and is not relevant anymore.

File: src/main/java/com/bio4j/angulillos/TypedLinkGraph.java (hunk 3 above)
Comment (eparejatobes) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: I'd rather use IllegalStateException, btw.

**Comment by eparejatobes** [comment]
https://github.com/bio4j/angulillos/pull/64#issuecomment-250773094 well, I don't see much point in `TypedSimpleGraph` in that case. What do you get from it? Why not just have the two classes everywhere and be done with it?

File: src/main/java/com/bio4j/angulillos/TypedLinkGraph.java (hunk 3 above)
Comment (laughedelic) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: I don't have a strong opinion on this, but `IllegalStateException` doesn't look related to this situation:

> Signals that a method has been invoked at an illegal or inappropriate time. In other words, the Java environment or Java application is not in an appropriate state for the requested operation.

**Comment by laughedelic** [comment]
The point of `TypedSimpleGraph` is to restict _all_ its edges to be only within that graph. Not a big deal, of course.

**Comment by eparejatobes** [comment]
Well if I'm reading your proposed design right it doesn't restrict anything, as you can use `GenericEdgeType` there

**Comment by laughedelic** [comment]
Oh, I forgot that making it `protected` won't restrict from using it in the second generation. But it could be `abstract`. Does it help? If there's no simple way to restrict it, then, of course, let's have just two classes.

File: src/main/java/com/bio4j/angulillos/TypedLinkGraph.java (hunk 3 above)
Comment (laughedelic) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: Maybe [`IntrospectionException`](https://docs.oracle.com/javase/8/docs/api/java/beans/IntrospectionException.html)?

**Comment by eparejatobes** [comment]
two classes plzz

File: src/main/java/com/bio4j/angulillos/TypedLinkGraph.java (hunk 3 above)
Comment (eparejatobes) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: IllegalState with cause IllegalArgument :yum:

Comment (laughedelic) [inline src/main/java/com/bio4j/angulillos/TypedLinkGraph.java]: `UnsupportedLookAndFeelException` also sound cool 👌

**Comment by laughedelic** [comment]
Done. So, if we allow any graph to have both kinds of edges, then we don't need any class hierarchy, just one class (`TypedGraph`) which can have generic edges between any other typed graphs and simple (internal) edges, that are just simple specializations of the generic ones.

**Comment by eparejatobes** [comment]
https://github.com/bio4j/angulillos/pull/64#issuecomment-250882201 fine.

**Comment by laughedelic** [comment]
LGTM
